### PR TITLE
Implement chat leaving flow

### PR DESCRIPTION
**Pull request by Vladyslav-Kuksiuk**
This PR adds an implementation of the chat leaving flow.

Chat leaving flow model:
image

### CODE REVIEW

File: server/src/main/java/io/spine/examples/chatspn/server/chat/ChatAggregate.java [hunk 1]
```diff
@@ -244,7 +250,7 @@ private ImmutableList<UserId> extractNewMembers(List<UserId> membersInCommand) {
      *
      * @throws ChatCannotBeMarkedAsDeleted
      *         if the user who told to delete personal chat wasn't a chat member,
-     *         or the user who told to delete group chat wasn't a chat owner,
+     *         or the user who told to delete group chat wasn't a chat owner / last member,
```
Comment (Artem-Semenov-dev) [inline server/src/main/java/io/spine/examples/chatspn/server/chat/ChatAggregate.java:253]: Please choose one style between '/' and 'or the'(it is below).

File: server/src/main/java/io/spine/examples/chatspn/server/chat/ChatAggregate.java [hunk 2]
```diff
@@ -290,6 +296,82 @@ private boolean checkDeletionPossibility(MarkChatAsDeleted c) {
...
+        if (state().getMemberList()
+                   .size() == 1) {
+            chatDeletionRequested = Optional.of(chatDeletionRequested(c));
+        }
+        return Pair.withOptional(userLeftChat, chatDeletionRequested);
+    }
+
+    @Apply
+    private void event(ChatDeletionRequested e) {
+    }
```
Comment (Artem-Semenov-dev) [inline server/src/main/java/io/spine/examples/chatspn/server/chat/ChatAggregate.java:331]: It looks weird to me. Please review.

File: server/src/test/java/io/spine/examples/chatspn/server/chat/ChatTest.java [hunk 3]
```diff
@@ -267,11 +276,85 @@ void rejectIfAlreadyMembers() {
...
+            ChatDeletionRequested chatDeletionRequested = chatDeletionRequestedFrom(command);
+            ChatDeleted chatDeleted = chatDeletedFrom(command);
+            UserLeftChat userLeftChat = userLeftChatFrom(command);
+
+            context().assertEvent(userLeftChat);
+            context().assertEvent(chatDeletionRequested);
+            context().assertEvent(chatDeleted);
+            context().assertEntity(chat.getId(), ChatAggregate.class)
+                    .deletedFlag()
+                    .isTrue();
```
Comment (Artem-Semenov-dev) [inline server/src/test/java/io/spine/examples/chatspn/server/chat/ChatTest.java:333]: Please adjust the code layout.

File: model/src/main/proto/spine_examples/chatspn/chat/chat.proto [hunk 4]
```diff
@@ -45,7 +45,7 @@ message Chat {
     ChatId id = 1;
 
     // List of the chat members.
-    repeated spine.core.UserId member = 2 [(required) = true, (distinct) = true];
+    repeated spine.core.UserId member = 2 [(distinct) = true];
```
Comment (Artem-Semenov-dev) [inline model/src/main/proto/spine_examples/chatspn/chat/chat.proto:48]: I think if the member's list can be empty It should be described in the documentation. Please review.

**Review by Artem-Semenov-dev — COMMENTED** [review]
@Vladyslav-Kuksiuk please see my comments.

File: server/src/test/java/io/spine/examples/chatspn/server/chat/given/ChatTestEnv.java [hunk 5]
```diff
@@ -295,4 +302,67 @@ public static MembersCannotBeRemoved membersCannotBeRemovedFrom(RemoveMembers co
                 .vBuild();
         return rejection;
     }
+
+    public static LeaveChat leaveChatCommand(Chat chat, UserId user) {
```
Comment (armiol) [inline server/src/test/java/io/spine/examples/chatspn/server/chat/given/ChatTestEnv.java:306]: Let's shorten to `leaveChat`.

**Review by armiol — CHANGES_REQUESTED** [review]
Please see my comments.

**Review by armiol — APPROVED** [review]
@Vladyslav-Kuksiuk LGTM with some comments to address.

**Pull request merged**